src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -80,10 +83,86 @@ await StoreWorkitemBlobAsync(identifier, dataset, cancellationToken)
             }
         }
 
+        public async Task CancelWorkitemAsync(string workitemInstanceUid, DicomDataset dataset, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(workitemInstanceUid, nameof(workitemInstanceUid));
+
+            try
+            {
+                var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
+
+                // TODO: Need to find a way to move this out of Orchestrator. Should be part of the main validator???
+                var futureProcedureStepState = await ValidateProcedureStepStateInStoreAsync(workitemInstanceUid, partitionKey, cancellationToken);
+
+                dataset.AddOrUpdate(DicomTag.ProcedureStepState, futureProcedureStepState);
+
+                var queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
+                var workitemKey = await _indexWorkitemStore
+                    .UpdateWorkitemAsync(partitionKey, workitemInstanceUid, dataset, queryTags, cancellationToken)
+                    .ConfigureAwait(false);
+
+                var workitemInstanceIdentifier = new WorkitemInstanceIdentifier(workitemInstanceUid, workitemKey, partitionKey);
+
+                dataset.TryGetString(DicomTag.ReasonForCancellation, out var cancellationReason);
+                if (!string.IsNullOrWhiteSpace(cancellationReason))
```

Comment:
We should only copy what we want not the entire dataset

Reply:
Done